File: flipt-client-java/src/test/java/TestFliptClient.java
import io.flipt.client.FliptClient;
import io.flipt.client.models.*;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class TestFliptClient {
  private static FliptClient fliptClient;

  @BeforeEach
  void init() {
    String fliptURL = System.getenv().get("FLIPT_URL");
    String clientToken = System.getenv().get("FLIPT_AUTH_TOKEN");

    assert fliptURL != null && !fliptURL.isEmpty();
    assert clientToken != null && !clientToken.isEmpty();
    fliptClient =
        FliptClient.builder()
            .url(fliptURL)
            .authentication(new ClientTokenAuthentication(clientToken))
            .build();
  }

  @Test
  void testNullFlagKey() throws Exception {
    Assertions.assertThrows(
        IllegalArgumentException.class,
        () -> {
          fliptClient.evaluateBoolean(null, "entity", new HashMap<>());
        });
  }

  @Test
  void testEmptyFlagKey() throws Exception {
    Assertions.assertThrows(
        IllegalArgumentException.class,
        () -> {
          fliptClient.evaluateBoolean("", "entity", new HashMap<>());
        });
  }

  @Test
  void testNullEntityId() throws Exception {
    Assertions.assertThrows(
        IllegalArgumentException.class,
        () -> {
          fliptClient.evaluateBoolean("flag1", null, new HashMap<>());
        });
  }

  @Test
  void testEmptyEntityId() throws Exception {
    Assertions.assertThrows(
        IllegalArgumentException.class,
        () -> {
          fliptClient.evaluateBoolean("flag1", "", new HashMap<>());
        });
  }

  @Test
  void testEvaluateVariant() throws Exception {
    Map<String, String> context = new HashMap<>();
    context.put("fizz", "buzz");

    VariantEvaluationResponse response = fliptClient.evaluateVariant("flag1", "entity", context);

    Assertions.assertEquals("flag1", response.getFlagKey());
    Assertions.assertTrue(response.isMatch());
    Assertions.assertEquals("MATCH_EVALUATION_REASON", response.getReason());
    Assertions.assertEquals("variant1", response.getVariantKey());
    Assertions.assertEquals("segment1", response.getSegmentKeys().get(0));
  }

  @Test
  void testEvaluateBoolean() throws Exception {
    Map<String, String> context = new HashMap<>();
    context.put("fizz", "buzz");

    BooleanEvaluationResponse response =
        fliptClient.evaluateBoolean("flag_boolean", "entity", context);

    Assertions.assertEquals("flag_boolean", response.getFlagKey());
    Assertions.assertTrue(response.isEnabled());
    Assertions.assertEquals("MATCH_EVALUATION_REASON", response.getReason());
  }

  @Test
  void testEvaluateBatch() throws Exception {
    Map<String, String> context = new HashMap<>();
    context.put("fizz", "buzz");

    List<EvaluationRequest> evalRequests =
        Arrays.asList(
            new EvaluationRequest("flag1", "entity", context),
            new EvaluationRequest("flag_boolean", "entity", context),
            new EvaluationRequest("notfound", "entity", context));

    BatchEvaluationResponse response = fliptClient.evaluateBatch(evalRequests);

    Assertions.assertEquals(3, response.getResponses().size());
    List<Response> responses = response.getResponses();

    Assertions.assertTrue(responses.get(0).getVariantEvaluationResponse().isPresent());
    VariantEvaluationResponse variantResponse =
        responses.get(0).getVariantEvaluationResponse().get();
    Assertions.assertEquals("flag1", variantResponse.getFlagKey());
    Assertions.assertTrue(variantResponse.isMatch());
    Assertions.assertEquals("MATCH_EVALUATION_REASON", variantResponse.getReason());
    Assertions.assertEquals("variant1", variantResponse.getVariantKey());
    Assertions.assertEquals("segment1", variantResponse.getSegmentKeys().get(0));

    Assertions.assertTrue(responses.get(1).getBooleanEvaluationResponse().isPresent());
    BooleanEvaluationResponse booleanResponse =
        responses.get(1).getBooleanEvaluationResponse().get();
    Assertions.assertEquals("flag_boolean", booleanResponse.getFlagKey());
    Assertions.assertTrue(booleanResponse.isEnabled());
    Assertions.assertEquals("MATCH_EVALUATION_REASON", booleanResponse.getReason());

    Assertions.assertTrue(responses.get(2).getErrorEvaluationResponse().isPresent());
    ErrorEvaluationResponse errorResponse = responses.get(2).getErrorEvaluationResponse().get();
    Assertions.assertEquals("notfound", errorResponse.getFlagKey());
    Assertions.assertEquals("default", errorResponse.getNamespaceKey());
    Assertions.assertEquals("NOT_FOUND_ERROR_EVALUATION_REASON", errorResponse.getReason());
  }

  @Test
  void testListFlags() throws Exception {
    List<Flag> flags = fliptClient.listFlags();
    Assertions.assertEquals(2, flags.size());
  }

  @Test
  void testGetSnapshot() throws Exception {
    String snapshot = fliptClient.getSnapshot();
    Assertions.assertNotNull(snapshot);
  }

  @Test
  void testSetGetSnapshotWithInvalidSnapshot() {
    FliptClient invalidFliptClient =
        FliptClient.builder()
            .url("http://localhost:8080")
            .errorStrategy(ErrorStrategy.FALLBACK)
            .snapshot("invalid")
            .build();
  }

  @Test
  void testSetGetSnapshotWithInvalidFliptURL() {
    // Get a snapshot from a working client
    String snapshot = fliptClient.getSnapshot();
    Assertions.assertNotNull(snapshot);

    // Create a client with the previous snapshot and an invalid URL
    FliptClient invalidFliptClient =
        FliptClient.builder()
            .url("http://invalid.flipt.com")
            .errorStrategy(ErrorStrategy.FALLBACK)
            .snapshot(snapshot)
            .build();

    Map<String, String> context = new HashMap<>();
    context.put("fizz", "buzz");

    for (int i = 0; i < 5; i++) {
      VariantEvaluationResponse response =
          invalidFliptClient.evaluateVariant("flag1", "entity", context);

      Assertions.assertEquals("flag1", response.getFlagKey());
      Assertions.assertTrue(response.isMatch());
      Assertions.assertEquals("MATCH_EVALUATION_REASON", response.getReason());
      Assertions.assertEquals("variant1", response.getVariantKey());
      Assertions.assertEquals("segment1", response.getSegmentKeys().get(0));

      BooleanEvaluationResponse booleanResponse =
          invalidFliptClient.evaluateBoolean("flag_boolean", "entity", context);

      Assertions.assertEquals("flag_boolean", booleanResponse.getFlagKey());
      Assertions.assertTrue(booleanResponse.isEnabled());
      Assertions.assertEquals("MATCH_EVALUATION_REASON", booleanResponse.getReason());

      List<Flag> flags = invalidFliptClient.listFlags();
      Assertions.assertEquals(2, flags.size());

      String snapshot = invalidFliptClient.getSnapshot();
      Assertions.assertNotNull(snapshot);
    }

    invalidFliptClient.close();
  }

  @AfterEach
  void tearDown() {
    if (fliptClient != null) fliptClient.close();
  }
}

File: flipt-client-kotlin-android/src/androidTest/java/io/flipt/client/TestFliptClient.kt
package io.flipt.client

import io.flipt.client.FliptClient
import io.flipt.client.models.*
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import java.io.File
import java.nio.charset.StandardCharsets
import java.util.Base64

class TestFliptClient {
    companion object {
        private val SNAPSHOT: String by lazy {
            Base64.getEncoder().encodeToString(readResourceFile("snapshot.json").toByteArray(StandardCharsets.UTF_8))
        }
        private val EMPTY_SNAPSHOT: String by lazy {
            Base64.getEncoder().encodeToString(readResourceFile("empty_snapshot.json").toByteArray(StandardCharsets.UTF_8))
        }

        private fun readResourceFile(filename: String): String {
            val resourcePath = "src/test/resources/$filename"
            return File(resourcePath).readText(Charsets.UTF_8)
        }
    }

    private var fliptClient: FliptClient? = null

    @Before
    @Throws(Exception::class)
    fun setUp() {
        val fliptURL = System.getenv("FLIPT_URL") ?: "http://10.0.2.2:8080"
        val clientToken = System.getenv("FLIPT_AUTH_TOKEN") ?: "secret"

        require(fliptURL.isNotEmpty()) { "FLIPT_URL environment variable is required" }
        require(clientToken.isNotEmpty()) { "FLIPT_AUTH_TOKEN environment variable is required" }
        require(fliptURL == "http://10.0.2.2:8080") { "FLIPT_URL must be http://10.0.2.2:8080" }

        fliptClient =
            FliptClient
                .builder()
                .url(url = fliptURL)
                .namespace("default")
                .authentication(ClientTokenAuthentication(clientToken))
                .build()
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        fliptClient?.close()
    }

    @Test
    @Throws(Exception::class)
    fun testEvaluateVariant() {
        val context: MutableMap<String, String> = HashMap()
        context["fizz"] = "buzz"

        val response = fliptClient?.evaluateVariant("flag1", "entity", context)

        assert("flag1" == response?.flagKey)
        assert(response?.match ?: false)
        assert("MATCH_EVALUATION_REASON" == response?.reason)
        assert("variant1" == response?.variantKey)
        assert("segment1" == response?.segmentKeys?.get(0))
    }

    @Test
    @Throws(Exception::class)
    fun testEvaluateBoolean() {
        val context: MutableMap<String, String> = HashMap()
        context["fizz"] = "buzz"

        val response = fliptClient?.evaluateBoolean("flag_boolean", "entity", context)

        assert("flag_boolean" == response?.flagKey)
        assert(response?.enabled ?: false)
        assert("MATCH_EVALUATION_REASON" == response?.reason)
    }

    @Test
    @Throws(Exception::class)
    fun testEvaluateBatch() {
        val context: MutableMap<String, String> = HashMap()
        context["fizz"] = "buzz"

        val evalRequests: Array<EvaluationRequest> =
            arrayOf(
                EvaluationRequest("flag1", "entity", context),
                EvaluationRequest("flag_boolean", "entity", context),
                EvaluationRequest("notfound", "entity", context),
            )

        val response = fliptClient?.evaluateBatch(evalRequests)

        assert(3 == response?.responses?.size)
        val responses = response?.responses

        assert(responses?.get(0)?.variantEvaluationResponse != null)
        val variantResponse = responses?.get(0)?.variantEvaluationResponse
        assert("flag1" == variantResponse?.flagKey)
        assert(variantResponse?.match ?: false)
        assert("MATCH_EVALUATION_REASON" == variantResponse?.reason)
        assert("variant1" == variantResponse?.variantKey)
        assert("segment1" == variantResponse?.segmentKeys?.get(0))

        assert(responses?.get(1)?.booleanEvaluationResponse != null)
        val booleanResponse = responses?.get(1)?.booleanEvaluationResponse
        assert("flag_boolean" == booleanResponse?.flagKey)
        assert(booleanResponse?.enabled ?: false)
        assert("MATCH_EVALUATION_REASON" == booleanResponse?.reason)

        assert(responses?.get(2)?.errorEvaluationResponse != null)
        val errorResponse = responses?.get(2)?.errorEvaluationResponse
        assert("notfound" == errorResponse?.flagKey)
        assert("default" == errorResponse?.namespaceKey)
        assert("NOT_FOUND_ERROR_EVALUATION_REASON" == errorResponse?.reason)
    }

    @Test
    @Throws(Exception::class)
    fun testListFlags() {
        val flags = fliptClient?.listFlags()
        assert(2 == flags?.size)
    }

    @Test
    fun testGetSnapshot() {
        val snapshot = fliptClient!!.getSnapshot()
        assertNotNull(snapshot)
        val expectedBytes = Base64.getDecoder().decode(SNAPSHOT)
        val actualBytes = Base64.getDecoder().decode(snapshot)
        val expectedJson = String(expectedBytes, StandardCharsets.UTF_8)
        val actualJson = String(actualBytes, StandardCharsets.UTF_8)
        // Use JSON comparison for leniency
        assertTrue(jsonEquals(expectedJson, actualJson))
    }

    @Test
    fun testSetGetSnapshotWithInvalidFliptURL() {
        val invalidFliptClient =
            FliptClient
                .builder()
                .url("http://invalid.flipt.com")
                .errorStrategy(ErrorStrategy.FALLBACK)
                .snapshot(SNAPSHOT)
                .build()

        val context: MutableMap<String, String> = HashMap()
        context["fizz"] = "buzz"

        repeat(5) {
            val response = invalidFliptClient.evaluateVariant("flag1", "entity", context)
            assert("flag1" == response.flagKey)
            assert(response.match)
            assert("MATCH_EVALUATION_REASON" == response.reason)
            assert("variant1" == response.variantKey)
            assert("segment1" == response.segmentKeys[0])

            val booleanResponse = invalidFliptClient.evaluateBoolean("flag_boolean", "entity", context)
            assert("flag_boolean" == booleanResponse.flagKey)
            assert(booleanResponse.enabled)
            assert("MATCH_EVALUATION_REASON" == booleanResponse.reason)

            val flags = invalidFliptClient.listFlags()
            assert(2 == flags.size)

            val snapshot = invalidFliptClient.getSnapshot()
            assertNotNull(snapshot)
        }

        invalidFliptClient.close()
    }

    // Simple JSON comparison ignoring order and whitespace
    private fun jsonEquals(
        expected: String,
        actual: String,
    ): Boolean {
        val expectedObj = com.fasterxml.jackson.module.kotlin.jacksonObjectMapper().readTree(expected)
        val actualObj = com.fasterxml.jackson.module.kotlin.jacksonObjectMapper().readTree(actual)
        return expectedObj == actualObj
    }
}
